File: demo/Fortran/2D/Sine_line/sine_line.f90
!* -------------------------------------------------------------------------- *
!* DESCRIPTION (reference phase where f(x,y) < 0):                            *
!* sinusoidal line in the square [0,1]x[0,1]                                  *
!* f(x,y) = y - B0*sin(C0 pi x+ pi/D0) - A0                                   *
!* -------------------------------------------------------------------------- *

SUBROUTINE CHECK_AREA(areanum)

  IMPLICIT NONE
  REAL(8), INTENT(IN) :: areanum

  REAL(8), PARAMETER :: MYPI = 3.141592653589793238462643D0
  REAL(8), PARAMETER :: A0 = 0.5D0, B0 = 0.25D0
  REAL(8), PARAMETER :: C0 = 4.0D0, D0 = 14.0D0

  REAL(8) :: areana

  INTRINSIC DCOS,DABS

  areana = A0 + B0*(-DCOS((C0 + 1./D0)*MYPI) + DCOS(MYPI/D0))/(C0*MYPI)

   write(*,*) '-----------------------------------------------------'
   write(*,*) '----------------- F: sine line check ----------------'
   write(*,*) ' '
   write(*,100) areana
   write(*,101) areanum
   write(*,102) DABS(areanum-areana)
   write(*,103) DABS(areanum-areana)/areana
   write(*,*) '-------------- F: end sine line check ---------------'
   write(*,*) '-----------------------------------------------------'
   write(*,*) ' '
   100 FORMAT('analytical area :', ES23.16)
   101 FORMAT('numerical area  :', ES23.16)
   102 FORMAT('absolute error  :', ES23.16)
   103 FORMAT('relative error  :', ES23.16)

  RETURN

END SUBROUTINE CHECK_AREA

!* -------------------------------------------------------------------------- *

REAL(8) FUNCTION IMPL_FUNC(xyz)

  IMPLICIT NONE

  REAL(8), DIMENSION(*), INTENT(IN) :: xyz

  REAL(8), PARAMETER :: MYPI = 3.141592653589793238462643D0
  REAL(8), PARAMETER :: A0 = 0.5D0, B0 = 0.25D0
  REAL(8), PARAMETER :: C0 = 4.0D0, D0 = 14.0D0

  REAL(8) :: x,y

  INTRINSIC DSIN

  x = xyz(1)
  y = xyz(2)
    
  IMPL_FUNC = y - A0 - B0*DSIN(C0*MYPI*x + MYPI/D0)

END FUNCTION IMPL_FUNC

File: demo/C/2D/Sine_line/sine_line.c
#include <stdio.h>
#include <math.h>
#include "sine_line.h"

typedef const double creal;
typedef const int cint;
typedef double real;

/* -------------------------------------------------------------------------- *
 * DESCRIPTION (reference phase where f(x,y) < 0):                            *
 * sinusoidal line in the square [0,1]x[0,1]                                  *
 * f(x,y) = y - B0*sin(C0 pi x+ pi/D0) - A0                                   *
 * -------------------------------------------------------------------------- */

int cont_line(real *xx,real *yy,cint ntot)
{
  int i;
  double dx,xi;

  dx = 1./(ntot-1);
  for (i=1; i<=ntot; i++) {
    xi = (i-1)*dx;
    xx[i] = xi;
    yy[i] = A0 + B0*sin(C0*MYPI*xi+MYPI/D0);
  }
  return ntot;
}
/* -------------------------------------------------------------------------- */

void check_area(creal area_n)
{
  real area_a;

  /* analytical integration with x in [0,1]  */

  area_a = A0 + B0*(-cos((C0 + 1./D0)*MYPI) + cos(MYPI/D0))/(C0*MYPI);

  fprintf (stderr,"-------------------------------------------------------\n");
  fprintf (stderr,"------------------ C: sine line check -----------------\n");
  fprintf (stderr,"analytical area : %23.16e\n",area_a);
  fprintf (stderr,"numerical  area : %23.16e\n\n",area_n);
  fprintf (stderr,"absolute error  : %23.16e\n",fabs(area_a-area_n));
  fprintf (stderr,"relative error  : %23.16e\n",fabs(area_a-area_n)/area_a); 
  fprintf (stderr,"-------------------------------------------------------\n");
  fprintf (stderr,"with Intel i7 3.4 GHz + Linux openSUSE 12.3 + gcc 4.7.2\n");
  fprintf (stderr,"-------------------------------------------------------\n");
  fprintf (stderr,"analytical area :  5.0000000000000000e-01\n");
  fprintf (stderr,"numerical  area :  4.9999999999993749e-01\n\n");
  fprintf (stderr,"absolute error  :  6.2505556286396313e-14\n");
  fprintf (stderr,"relative error  :  1.2501111257279263e-13\n");
  fprintf (stderr,"--------------- C: end sine line check ----------------\n");
  fprintf (stderr,"-------------------------------------------------------\n");

  return;
}
/* -------------------------------------------------------------------------- */

real impl_func(creal xy[]) 
{
  double x,y,f0;

  x = xy[0];
  y = xy[1];

  f0 = y - A0 - B0*sin(C0*MYPI*x + MYPI/D0);

  return f0;
}
